File: app/src/main/java/com/taskmanager/ui/TaskBarView.java
package com.taskmanager.ui;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.util.AttributeSet;
import android.view.View;

import java.util.ArrayList;

/**
 * TODO: document your custom view class.
 */
public class TaskBarView extends View {
    Paint paint;
    ArrayList<TaskBarItem> items;

    public TaskBarView(Context context) {
        super(context);
        init(null, 0);
    }

    public TaskBarView(Context context, AttributeSet attrs) {
        super(context, attrs);
        init(attrs, 0);
    }

    public TaskBarView(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
        init(attrs, defStyle);
    }

    private void init(AttributeSet attrs, int defStyle) {
        // Load attributes
        /*
        final TypedArray a = getContext().obtainStyledAttributes(
                attrs, R.styleable.TaskBarView, defStyle, 0);

        a.recycle();
        */

        paint = new Paint();
        items = new ArrayList<TaskBarItem>();
    }

    public void clearItems() {
        items.clear();
        postInvalidate();
    }

    public void addItem(TaskBarItem item) {
        items.add(item);
        postInvalidate();
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);

        // TODO: consider storing these as member variables to reduce
        // allocations per draw cycle.
        int paddingLeft = getPaddingLeft();
        int paddingTop = getPaddingTop();
        int paddingRight = getPaddingRight();
        int paddingBottom = getPaddingBottom();

        int contentWidth = getWidth() - paddingLeft - paddingRight;
        int contentHeight = getHeight() - paddingTop - paddingBottom;

        paint.setStyle(Paint.Style.STROKE);
        paint.setColor(Color.BLACK);
        paint.setStrokeWidth(4);
        canvas.drawRect(paddingLeft,paddingTop,paddingLeft + contentWidth, paddingTop + contentHeight, paint);

        int left = paddingLeft;
        for (TaskBarItem item : items) {
            int len = (int)((float)contentWidth * item.length);

            paint.setStyle(Paint.Style.FILL);
            paint.setColor(item.color);
            canvas.drawRect(left, paddingTop, left + len, paddingTop + contentHeight, paint);

            if (item.highlight) {
                paint.setStyle(Paint.Style.STROKE);
                paint.setColor(Color.RED);
                canvas.drawRect(left, paddingTop, left + len, paddingTop + contentHeight, paint);
            }

            left += len;
        }
    }

    public static class TaskBarItem {
        int type;
        float length;
        int color;
        boolean highlight;

        public TaskBarItem(int type, float length, int color, boolean highlight) {
            this.type = type;
            this.length = length;
            this.color = color;
            this.highlight = highlight;
        }
    }
}

File: taskmanager/src/main/java/com/example/alexeyglushkov/taskmanager/ui/TaskBarView.kt
package com.example.alexeyglushkov.taskmanager.ui

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.util.AttributeSet
import android.view.View

import java.util.ArrayList

/**
 * TODO: document your custom view class.
 */
class TaskBarView : View {
    protected lateinit var paint: Paint
    protected lateinit var items: ArrayList<TaskBarItem>

    constructor(context: Context) : super(context) {
        init(null, 0)
    }

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {
        init(attrs, 0)
    }

    constructor(context: Context, attrs: AttributeSet, defStyle: Int) : super(context, attrs, defStyle) {
        init(attrs, defStyle)
    }

    private fun init(attrs: AttributeSet?, defStyle: Int) {
        // Load attributes
        /*
        final TypedArray a = getContext().obtainStyledAttributes(
                attrs, R.styleable.TaskBarView, defStyle, 0);

        a.recycle();
        */

        paint = Paint()
        items = ArrayList()
    }

    fun clearItems() {
        items.clear()
        postInvalidate()
    }

    fun addItem(item: TaskBarItem) {
        items.add(item)
        postInvalidate()
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)

        // TODO: consider storing these as member variables to reduce
        // allocations per draw cycle.
        val paddingLeft = paddingLeft
        val paddingTop = paddingTop
        val paddingRight = paddingRight
        val paddingBottom = paddingBottom

        val contentWidth = width - paddingLeft - paddingRight
        val contentHeight = height - paddingTop - paddingBottom

        paint.style = Paint.Style.STROKE
        paint.color = Color.BLACK
        paint.strokeWidth = 4f
        canvas.drawRect(paddingLeft.toFloat(), paddingTop.toFloat(), (paddingLeft + contentWidth).toFloat(), (paddingTop + contentHeight).toFloat(), paint)

        var left = paddingLeft
        for (item in items) {
            val len = (contentWidth.toFloat() * item.length).toInt()

            paint.style = Paint.Style.FILL
            paint.color = item.color
            canvas.drawRect(left.toFloat(), paddingTop.toFloat(), (left + len).toFloat(), (paddingTop + contentHeight).toFloat(), paint)

            if (item.highlight) {
                paint.style = Paint.Style.STROKE
                paint.color = Color.RED
                canvas.drawRect(left.toFloat(), paddingTop.toFloat(), (left + len).toFloat(), (paddingTop + contentHeight).toFloat(), paint)
            }

            left += len
        }
    }

    class TaskBarItem(internal var type: Int, internal var length: Float, internal var color: Int, internal var highlight: Boolean)
}
